src/scene/procedural.js

```diff
@@ -17,192 +17,295 @@ const primitiveUv1PaddingScale = 1.0 - primitiveUv1Padding * 2;
 // cached mesh primitives
 const shapePrimitives = [];
 
+/**
+ * Normalizes the entire array in place.
+ * 
+ * @example ```js
+ *   let arr = [0,10,0, 20,20,20, 0,1,30];
+ *   normalizeArray(arr);
+ *   arr.map(_ => _.toFixed(3));
+ *   // Output: ['0.000', '1.000', '0.000', '0.577', '0.577', '0.577', '0.000', '0.033', '0.999']
+ * ```
+ * @param {number[] | Float32Array} arr
+ */
+function normalizeArray(arr) {
+    const n = arr.length;
+    const sqrt = Math.sqrt;
+    for (let i = 0; i < n; i += 3) {
+        const nx = arr[i];
+        const ny = arr[i + 1];
+        const nz = arr[i + 2];
+        const invLen = 1 / sqrt(nx * nx + ny * ny + nz * nz);
+        arr[i] *= invLen;
+        arr[i + 1] *= invLen;
+        arr[i + 2] *= invLen;
+    }
+}
+
 /**
  * Generates normal information from the specified positions and triangle indices. See
  * {@link createMesh}.
  *
- * @param {number[]} positions - An array of 3-dimensional vertex positions.
- * @param {number[]} indices - An array of triangle indices.
- * @returns {number[]} An array of 3-dimensional vertex normals.
+ * @param {number[]|Uint16Array|Float32Array|Int8Array|Uint8Array|Int16Array|Int32Array|Uint32Array} positions - An array of 3-dimensional vertex positions.
+ * @param {number[]|Uint16Array|Float32Array|Int8Array|Uint8Array|Int16Array|Int32Array|Uint32Array} indices - An array of triangle indices.
+ * @returns {Float32Array} An array of 3-dimensional vertex normals.
  * @example
  * var normals = pc.calculateNormals(positions, indices);
  * var tangents = pc.calculateTangents(positions, normals, uvs, indices);
  * var mesh = pc.createMesh(positions, normals, tangents, uvs, indices);
  */
 function calculateNormals(positions, indices) {
```

Comment:
Do we use IIFE to store temporal variables in the engine to be garbage-collector-friendly?

For example,

```js
const calculateNormals = (() => {
  const vec3Tmp = new pc.Vec3();
  const vec3Tmp2 = new pc.Vec3();
  
  const quatTmp = new pc.Quat();
  
  return function calculateNormals(positions, indices) {
    // here is the main code
    const p1p2 = vec3Tmp;
    const p1p3 = vec3Tmp2;
    // lalilulelo
   }
})()
```

Reply:
Hey @querielo, nice and useful idea, we used to have that style for ES5 prototype style methods, but ES6 doesn't support it (sadly). For a normal function like this it still works of course.

Why do you do `const p1p2 = vec3Tmp` tho, don't you like closure variables? Or any problem with just naming the temporary variable `p1p2` in the first place?

Reply:
@kungfooman

> but ES6 doesn't support it (sadly).

Could you explain what do you mean?

The next pattern works well:
```js
class Foo {
    bar = (() => {
        const vec3Tmp = new pc.Vec3();

        return function bar() {
             // work with vec3Tmp
        }
    })()
}
```

So, here we store temporal variable in closure that is created by IIFE.

Pros:
1) garbage-collection-freindly
2) `vec3Tmp` is defined as close to `bar` as possible. Less probability that somebody else reuses `vec3Tmp` and can break other function

Cons:
1) It's hard to read if you don't know the pattern.
2) A new instance of Foo create a new function **bar**.
3) `bar` is stored in instance of Foo. It cannot be overridden or extended by subclasses as usual methods. The method can be overridden in the next way:

```js
class Bar extends Foo {
    bar = (() => {
        const barFromFoo = this.bar

        return function bar() {
            barFromFoo()

            // write your extension
        }
    })()
}
```

**NOTE: If we are talking about functions (not methods) then the second and third cons are not valid.**

I think, it may be more appropriate to open a new issue to address this topic, as it seems to be unrelated to the current pull request.

Reply:
Well, usually it was just:

```js
function Test() {
    // nothing yet
}
Test.prototype.hai = (() => {
    const oldNow = Date.now();
    return function bar() {
        console.log("Test#hai> oldNow", oldNow)
    }
})();
const test = new Test();
test.hai();
```

But the ES6 rewrite can't handle that, so the temporary variables just went into the local file scope (or into static class variables in some other cases [1]).

[1] https://github.com/playcanvas/engine/blob/main/src/framework/anim/evaluator/anim-target-value.js#L17

I basically agree with your pros/cons.

Syntax that comes closest to ES6 might be:

```js
function Test() {
    // nothing yet
}
Object.assign(Test.prototype, {
    hai: (() => {
        const oldNow = Date.now();
        return function bar() {
            console.log("Test#hai> oldNow", oldNow)
        }
    })(),
    
    bye() {
        console.log("bye");
    }
});
const test = new Test();
test.hai();
test.bye();
```

But rewriting everything just to replace some tmp vectors in local file scope via IIFE methods is kind of overkill for classes (for functions I still think it's a good idea, as you proposed).

Do you want to answer my initial two questions in previous post about closures?

Reply:
@kungfooman - any update on this PR or should we close it for now?